components/profile/heartbeats.rs

```diff
@@ -12,39 +12,67 @@ use std::env::var_os;
 use std::error::Error;
 use std::fs::File;
 use std::path::Path;
+use std::sync::atomic::{ATOMIC_BOOL_INIT, AtomicBool, Ordering};
 
 
 static mut HBS: Option<*mut HashMap<ProfilerCategory, Heartbeat>> = None;
 
+// unfortunately can't encompass the actual hashmap in a Mutex (Heartbeat isn't Send/Sync), so we'll use a spinlock
+static mut HBS_SPINLOCK: AtomicBool = ATOMIC_BOOL_INIT;
+
+fn lock_and_work<F, R>(work: F) -> R where F: FnOnce() -> R {
+    unsafe {
+        while HBS_SPINLOCK.compare_and_swap(false, true, Ordering::SeqCst) {}
+    }
+    let result = work();
+    unsafe {
+        HBS_SPINLOCK.store(false, Ordering::SeqCst);
+    }
+    result
+}
+
 /// Initialize heartbeats
 pub fn init() {
     let mut hbs: Box<HashMap<ProfilerCategory, Heartbeat>> = Box::new(HashMap::new());
     maybe_create_heartbeat(&mut hbs, ProfilerCategory::ApplicationHeartbeat);
-    unsafe {
-        HBS = Some(Box::into_raw(hbs));
-    }
+    lock_and_work(||
+        unsafe {
+            HBS = Some(Box::into_raw(hbs));
```

Comment:
Can we safely touch HBS like that outside of the locking?  Also, are you talking about doing this everywhere we use `lock_and_work`?  My Rust is apparently really rusty - would I need to first define the closure outside of the call to `lock_and_work` in order to pass `&mut HBS` to it?  Thanks.